pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -212,82 +292,37 @@ func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) [
-func (state PipelineRunState) SuccessfulOrSkippedDAGTasks(d *dag.Graph) []string {
+func (rpr ResolvedPipelineRun) successfulOrSkipped() []string {
 	tasks := []string{}
-	stateMap := state.ToMap()
-	for _, t := range state {
-		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
-			if t.IsSuccessful() || isSkipped(t, stateMap, d) {
-				tasks = append(tasks, t.PipelineTask.Name)
-			}
+	for _, t := range rpr.DagState {
+		if t.IsSuccessful() || rpr.IsTaskSkipped(t.PipelineTask.Name) {
+			tasks = append(tasks, t.PipelineTask.Name)
 		}
 	}
 	return tasks
 }
 
-// isDAGTasksStopped returns true if any of the pipeline task has failed
-// and none of the pipeline task are still running
-func (state PipelineRunState) isDAGTasksStopped(d *dag.Graph) bool {
-	failed := false
-	for _, t := range state {
-		if t.IsFailure() {
-			failed = true
-			continue
-		}
-		if t.IsStarted() && !t.IsDone() {
-			failed = false
-			break
-		}
-	}
-	return failed
-}
-
-// checkTasksDone returns true if all tasks from the specified graph are finished executing
-// a task is considered done if it has failed/succeeded/skipped
-func (state PipelineRunState) checkTasksDone(d *dag.Graph) (isDone bool) {
-	isDone = true
-	stateMap := state.ToMap()
-	for _, t := range state {
-		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
-			if t.TaskRun == nil {
-				// this task might have skipped if taskRun is nil
-				// continue and ignore if this task was skipped
-				// skipped task is considered part of done
-				if isSkipped(t, stateMap, d) {
-					continue
-				}
-				return false
-			}
-			isDone = isDone && t.IsDone()
-			if !isDone {
-				return
-			}
-		}
-	}
-	return
-}
-
 // GetFinalTasks returns a list of final tasks without any taskRun associated with it
 // GetFinalTasks returns final tasks only when all DAG tasks have finished executing successfully or skipped or
 // any one DAG task resulted in failure
-func (state PipelineRunState) GetFinalTasks(d *dag.Graph, dfinally *dag.Graph) []*ResolvedPipelineRunTask {
+func (rpr ResolvedPipelineRun) GetFinalTasks() []*ResolvedPipelineRunTask {
 	tasks := []*ResolvedPipelineRunTask{}
-	finalCandidates := map[string]struct{}{}
-	// check either pipeline has finished executing all DAG pipelineTasks
-	// or any one of the DAG pipelineTask has failed
-	if state.isDAGTasksStopped(d) || state.checkTasksDone(d) {
-		// return list of tasks with all final tasks
-		for _, t := range state {
-			if _, ok := dfinally.Nodes[t.PipelineTask.Name]; ok && !t.IsSuccessful() {
-				finalCandidates[t.PipelineTask.Name] = struct{}{}
-			}
+	// If the main DAG is still running, return nothing
+	if !rpr.isDAGDone() {
+		return tasks
+	}
+	// return list of tasks with all final tasks
+	for _, t := range rpr.FinalState {
+		// Final states cannot be skipped, so it's enough to check is a Task is not started
+		// and not done in case of failure
+		if t.TaskRun == nil {
+			tasks = append(tasks, t)
 		}
-		tasks = state.GetNextTasks(finalCandidates)
```

Comment:
I haven't reviewed all the changes yet so I might be wrong but wanted to make sure we are covering cases where a final task fails and need to reschedule for another retry. Just checking `TaskRun` not nil here is not sufficient. If a final task has failed, it will not be rescheduled even if the retries are still left, please correct me if I am wrong 🙏

Reply:
You're right, the failed retries needs to be covered for finally too. 
If that is something we want to have - which I think it's the case - I'll add test coverage for it and make sure it's covered here.

Reply:
yes we do if we don't have tests for them, please add or let me know if you want me to add them 🙏

Reply:
Since we have removed `GetNextTasks` from here, you will have to add separate tests for final tasks. I still think it would be easier to create a common function `getNextTasks` which we can unit test irrespective of DAG tasks or final tasks. `getNextTasks` can be used by both `GetNextTasks` and `GetFinalTasks`.